### ISSUE TITLE
multi player exception

### ISSUE THREAD

**Issue by gitlsl** [ZeBobo5/Vlc.DotNet#430]
I have an issue / a question (pick one) about Vlc.DotNet.

# Generic information

- **Vlc.DotNet version** : (Please fill)  2.2.1
- **Vlc.DotNet project used** : (WinForms/WPF/Core)       console
- **libvlc version** : (Please fill) (x86/x64) + how did you get it?  VideoLAN.LibVLC.Windows 3.0.0-alpha3 from nuget
- **.net version** : (Please fill)  4.6.1
- **Project language** : (C#/VB.net) c#
- **Project build architecture** : (x86/x64/AnyCPU/AnyCPU (prefer 32bits))   x86
- **Operating system** : (Windows ...) (x86/x64)  win10 x64

# Summary
I have many flv,rtmp urls, just want to check them if was alive
to improve speed, I use multithread
`
  CancellationTokenSource s = new CancellationTokenSource();
          

            var mediaPlayer = GetVlcMediaPlayer();

            var mediaOptions = new string[]
            {
                 ":novideo",
                ":noaudio"
            };

            mediaPlayer.SetMedia(new Uri(url), mediaOptions);

        
            mediaPlayer.Playing += OnMediaPlayerOnPlaying;
            void OnMediaPlayerOnPlaying(object sender, VlcMediaPlayerPlayingEventArgs eventArgs)
            {
                mediaPlayer.Playing -= OnMediaPlayerOnPlaying;
                s.Cancel();
              
            }
            mediaPlayer.Play();
            if (s.Token.WaitHandle.WaitOne(delay * 1000))
            {
                mediaPlayer.Stop();
                mediaPlayer.Dispose();
                return true;
            }

            Console.WriteLine(mediaPlayer.State);
            mediaPlayer.Stop();
            mediaPlayer.Dispose();
            return false;
        }
`
 single thread test:
if mediaPlayer was Singleton ，it will work ok
if mediaPlayer  was   local in method , after 30+ play it will get exception (Managed Debugging Assistants LoaderLock), break at 

`
    mediaPlayer = new Vlc.DotNet.Core.VlcMediaPlayer(libDirectory);
`

multithread test:
if mediaPlayer  was   local in method ,it will get exception (key not find in param) soon,  also  break at 

`
    mediaPlayer = new Vlc.DotNet.Core.VlcMediaPlayer(libDirectory);
`

**Comment by jeremyVignelles** [comment]
Can it be related to #397 ?

**Comment by gitlsl** [comment]
I think multithread test is related to #397  single thread test: don;t related
this can be easy reproduce also

**Comment by jeremyVignelles** [comment]
Please try with the latest develop package.

I have to admit that I don't understand your issue. What is the exception?

**Comment by jeremyVignelles** [comment]
After reading again, it looks like you are calling `s.Cancel()` which unlocks the `WaitOne`.
My guess is that Cancel would call the rest of your code in a synchronous way, thus causing a situation similar to https://github.com/ZeBobo5/Vlc.DotNet/wiki/Vlc.DotNet-freezes-(don't-call-Vlc.DotNet-from-a-Vlc.DotNet-callback)

Can you check if the solution provided there can help? (Switch thread to call `s.Cancel()`)

**Comment by gitlsl** [comment]
@JeanMeijer    I follow your solution ,  no (Managed Debugging Assistants LoaderLock) exception   any time ,thank you 
but in multithread  situation  ( thread count > 5 )  it will get exception (System.Collections.Generic.KeyNotFoundException)   on
`mediaPlayer = new Vlc.DotNet.Core.VlcMediaPlayer(libDirectory);`

**Comment by gitlsl** [comment]
````
	internal VlcMedia(VlcMediaPlayer player, VlcMediaInstance mediaInstance)
		{
			if (!VlcMedia.LoadedMedias.ContainsKey(player))
			{
				VlcMedia.LoadedMedias[player] = new List<VlcMedia>();
			}
			VlcMedia.LoadedMedias[player].Add(this);
			this.MediaInstance = mediaInstance;
			this.myVlcMediaPlayer = player;
			this.RegisterEvents();
		}
````

		public static VlcManager GetInstance(DirectoryInfo dynamicLinkLibrariesPath)
		{
			if (!VlcManager.myAllInstance.ContainsKey(dynamicLinkLibrariesPath))
			{
				VlcManager.myAllInstance[dynamicLinkLibrariesPath] = new VlcManager(dynamicLinkLibrariesPath);
			}
			return VlcManager.myAllInstance[dynamicLinkLibrariesPath];
		}
`

I guess above code  is  not thread_safe  ,so we will get above exception

**Comment by jeremyVignelles** [comment]
You're probably right. You can make the PR if you want

**Comment by dilandau2001** [comment]
I have tried this "fix" and it totally works.
I just needed to use a lock for LoadedMedias resource.

In a multithread/multiplayer environment it could happen that this line is called at the same time by multiple players:
VlcMedia.LoadedMedias[player] = new List<VlcMedia>();

Although it is giving a nullreference exception in the end the original error is due to accessing a modified collection.

I first thought it should be usinc a ConcurrentDictionary but this is not supported in .Net20 so the lock is the only "fast" solution. 

Unfortunatelly I have no access to github clone repositories from my location to make the PR.
